File: src/pages/api/crawl/processNftImages.js
// pages/api/processNftImages.js
import db from "../../../lib/db";
import uploadFileToSpaces from "./uploadFileToSpaces"; 

export default async function handler(req, res) {
    if (req.method !== 'GET') {
        return res.status(405).end("Method Not Allowed");
    }

    try {
        const nftsToProcess = await db.any(`
            SELECT * FROM nfts WHERE media_url NOT LIKE 'https://shuk.nyc3.cdn.digitaloceanspaces.com/%'
        `);

        let processedCount = 0;
        const maxProcessCount = 10000; // Limit to 5000 NFTs to avoid timeouts

        for (const nft of nftsToProcess) {
            if (processedCount >= maxProcessCount) {
                break; // Stop processing if the limit is reached
            }

            const newImageUrl = await uploadFileToSpaces(nft.media_url, `${nft.contract_address}/${nft.token_id}`);
            await db.none(`
                UPDATE nfts SET media_url = $1 WHERE nft_id = $2
            `, [newImageUrl, nft.nft_id]);

            processedCount++; // Increment the counter after each successful process
        }

        res.status(200).json({ message: `Processed ${processedCount} NFT images successfully.` });
    } catch (error) {
        console.error("Failed to process NFT images:", error);
        res.status(500).json({ error: "Failed to process NFT images" });
    }
}

File: database-toolkit/actions/cleanup/process_nft_images.py
import requests
import time
from db.connection import connect_db
from psycopg2.extras import DictCursor
from s3.queries import upload_file_to_s3

def process_nft_images():
    conn = connect_db()
    if not conn:
        print("Failed to connect to the database.")
        return

    cursor = conn.cursor(cursor_factory=DictCursor)
    try:
        print("Fetching NFTs with missing media URLs...")
        cursor.execute("""
            SELECT * FROM nfts WHERE media_url NOT LIKE 'https://shuk.nyc3.cdn.digitaloceanspaces.com/%'
        """)
        nfts_to_process = cursor.fetchall()
        print(f"Found {len(nfts_to_process)} NFTs with missing media URLs.")

        processed_count = 0
        max_process_count = 100000  

        for nft in nfts_to_process:
            if processed_count >= max_process_count:
                print(f"Reached the maximum process count of {max_process_count}.")
                break  # Stop processing if the limit is reached

            new_image_url = upload_file_to_s3(nft['media_url'], f"{nft['contract_address']}/{nft['token_id']}", "shuk")
            cursor.execute("""
                UPDATE nfts SET media_url = %s WHERE nft_id = %s
            """, (new_image_url, nft['nft_id']))
            conn.commit()
            print(f"Updated NFT ID {nft['nft_id']} with new media URL: {new_image_url}")
            processed_count += 1

        print(f"Processed {processed_count} NFT images successfully.")
    except Exception as error:
        print(f"Failed to process NFT images: {error}")
        conn.rollback()
    finally:
        cursor.close()
        conn.close()

if __name__ == "__main__":
    process_nft_images()
